src/main/scala/com/fulcrumgenomics/fastq/DemuxFastqs.scala

```diff
@@ -0,0 +1,614 @@
+  /** A class to store the sample ordinal and associated demultiplexed [[SAMRecord]]s.
+    * @param sampleIndex the sample index to which the records were assigned (0-based)
+    * @param numMismatches the # of mismatches if it matched a sample with a sample barcode, -1 otherwise (the 'unmatched'
+    *                      sample).
+    * @param records the records, one for each read that have template bases.
+    */
+  case class DemuxRecord(sampleIndex: Int, numMismatches: Int, records: Seq[SAMRecord])
+
+  /** Counts the nucleotide mismatches between two strings of the same length.  Ignores no calls in expectedBases.
+    * Observed base qualities less than the minimum base quality are counted as mismatches if not a no call.
+    */
+  private[fastq] def countMismatches(observedBases: Array[Byte], expectedBases: Array[Byte]): Int = {
+    var idx = 0
+    var count = 0
+    while (idx < observedBases.length) {
+      val expectedBase = expectedBases(idx)
+      val observedBase = observedBases(idx)
+      if (!SequenceUtil.isNoCall(expectedBase) && !SequenceUtil.basesEqual(observedBase, expectedBase)) {
+        count += 1
+      }
+      idx += 1
+    }
+    count
+  }
+}
+
+/** Performs sample demultiplexes from reads from the same template/fragment.
+  *
+  * A [[SAMFileHeader]] should be given per sample and a [[ReadStructure]] per read from the same template/fragment.
+  * Use the [[demultiplex()]] method to create a [[SAMRecord]] for each read with template bases.  Any molecular barcodes
+  * will be extracted and stored in the tag specified by [[umiTag]].
+  *
+  * @param headers a sequence of [[SAMFileHeader]]s for each sample in the same order as samples in the sample sheet.
+  * @param samples the samples ordered by sample ordinal.
+  * @param readStructures the read structures, one for each read that will be given to [[demultiplex()]].
+  * @param umiTag the tag to store any molecular barcodes.  The barcodes from reads will be delimited by "-".
+  * @param qualityFormat the quality format for the FASTQ records.
+  * @param minQ the minimum base quality to enforce.
+  * @param maxQ the maximum base quality to enforce.
+  * @param maxMismatches the maximum mismatches to match a sample barcode.
+  * @param minMismatchDelta the minimum difference between number of mismatches in the best and second best barcodes for
+  *                         a barcode to be considered a match.
+  * @param maxNoCalls the maximum number of no calls in the sample barcode bases allowed for matching.
+  */
+private class FastqDemultiplexer(val headers: Seq[SAMFileHeader],
+                                 val samples: Seq[Sample],
+                                 val readStructures: Seq[ReadStructure],
+                                 val umiTag: String = "RX",
+                                 val qualityFormat: FastqQualityFormat = FastqQualityFormat.Standard,
+                                 val minQ: Int = 0,
```

Comment:
I feel like enforcement of the min/max quality doesn't belong here.  This is really validation that the fastqs aren't borked, and surely that should reside with the thing reading the fastqs?  Perhaps we should add options to FastqSource to translate qualities and enforce the quality range, so that it's re-useable elsewhere?